### DISCUSSION TITLE
MAUI Binding .aar library

### DISCUSSION THREAD

**Discussion by vfongmala** [dotnet/maui#1340] [Q&A]
I tried to adapt the approach from https://docs.microsoft.com/en-us/xamarin/android/platform/binding-java-library/binding-an-aar into my MAUI project.
But I have not succeeded so far. I cannot reference the binding project into the main project.
Any suggestion on how can I use native library in MAUI project?

**Comment by charleszh-dev** [comment]
In .net6.0, there is a new project template Android Java Library Binding (android-bindinglib), you can use it to create your binding library. However, the binding library of xamarin can also be used in maui, so I think maybe you encountered a problem when creating the binding library.

**Reply by charleszh-dev** [reply]
I'm also trying to create a binding library, but I'm having problems, this is the relevant [issue](https://github.com/dotnet/maui/issues/3245). do you know how to solve this problem? Thanks.

**Reply by AdamDiament** [reply]
Hi @realZhangChi -
re

> However, the binding library of xamarin can also be used in maui

This sounds great! Is this documented anywhere?

**Reply by ksivasenthil** [reply]
> However, the binding library of xamarin can also be used in maui, so I think maybe you encountered a problem when creating the binding library.

This might not be entirely true. Though the official document (@AdamDiament I was also searching for this!) [MAUI migration guide for binding libraries](https://github.com/xamarin/xamarin-android/blob/main/Documentation/guides/OneDotNetBindingProjects.md) calims the following - 

  > In .NET 6+, there is no longer a concept of a [binding project](https://docs.microsoft.com/xamarin/android/platform/binding-java-library/) as a separate project type. **Any of the MSBuild item groups or build actions** that currently work in binding projects will be supported through a .NET 6+ Android application or library.

Despite that we were having trouble making a Android Bindings Library (Xamarin) compile in Azure DevOps CI pipeline environment post MAUI migration for the app that consumed the Android Binding Library (Xamarin).

We could not get past the step `_AndroidAotCompilation target` with relevant error message being the following

``` log
error : Mono: Request to load PdfViewerX in alc 000001F7A9848DB0 [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
error : Mono: The following assembly referenced from D:\a\1\s\---\---\obj\Release\net7.0-android\android-x64\linked\Core.dll could not be loaded: [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
error :      Assembly:   PdfViewerX    (assemblyref_index=21) [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
error :      Version:    1.0.0.0 [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
error : Mono: Failed to load assembly Core[000001F7A9845EC0]. [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
error : Failed to load method 0x60002b0 from 'D:\a\1\s\---\---\obj\Release\net7.0-android\android-x64\linked\Core.dll' due to Could not load file or assembly 'PdfViewerX, Version=1.0.0.0, Culture=neutral, PublicKeyToken=null' or one of its dependencies.. [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
error : AOT of image obj\Release\net7.0-android\android-x64\linked\Core.dll failed. [D:\a\1\s\---\---\---.csproj::TargetFramework=net7.0-android]
```

Needless to say, many steps above this the PdfViewerX was built successfully and the binding project has the aar embedded inside the PdfViewerX.dll using the LibraryProjectZip build action.

Please let me know if we are doing something silly but we are going to try by switching to the SDK style project available from the MAUI Android Java Library Binding.

**Comment by rkops-bd** [comment]
I don't have the buildaction LibraryProjectZip when trying to create a binding for an .aar file in VS2022 Preview

**Reply by Kebechet** [reply]
use `AndroidLibrary` instead

**Comment by yushulx** [comment]
In Visual Studio 2022 preview, create an `Android Java Library Binding` project.

![image](https://user-images.githubusercontent.com/2202306/185534745-c081da9a-3eac-4492-adb2-124bf94c6d91.png)

Then click the aar file and change the build action to `AndroidLibrary`

![image](https://user-images.githubusercontent.com/2202306/185534845-0ca9f394-a5d9-482e-96e6-35a9aee7a502.png)

Here is my project: https://github.com/yushulx/barcode-dotnet-mobile/tree/main/android

**Comment by rbprakash** [comment]
i',m trying to Binding an .arr file by following above mentioned steps. getting these kind of errors, can some one help ?

Severity	Code	Description	Project	File	Line	Suppression State
Error	CS1001	Identifier expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS1513	} expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS1514	{ expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	10	Active
Error	CS1026	) expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	11	Active
Error	CS1031	Type expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	11	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	11	Active
Error	CS1002	; expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	11	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	11	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	35	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	35	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	35	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	35	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	35	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	36	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS8803	Top-level statements must precede namespace and type declarations.	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1525	Invalid expression term 'int'	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1002	; expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1026	) expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1002	; expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	41	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	58	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:\MAUI\AndroidBinding1\obj\Debug	59	Active
Error	CS1001	Identifier expected	AndroidBinding1	C:	10	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:	10	Active
Error	CS1513	} expected	AndroidBinding1	C:	10	Active
Error	CS1514	{ expected	AndroidBinding1	C:	10	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	10	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:	10	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	10	Active
Error	CS1026	) expected	AndroidBinding1	C:	11	Active
Error	CS1031	Type expected	AndroidBinding1	C:	11	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:	11	Active
Error	CS1002	; expected	AndroidBinding1	C:	11	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	11	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:	35	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:	35	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:	35	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	35	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:	35	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	36	Active
Error	CS0116	A namespace cannot directly contain members such as fields, methods or statements	AndroidBinding1	C:	41	Active
Error	CS1056	Unexpected character 'Ɪ'	AndroidBinding1	C:	41	Active
Error	CS8803	Top-level statements must precede namespace and type declarations.	AndroidBinding1	C:	41	Active
Error	CS1525	Invalid expression term ')'	AndroidBinding1	C:	41	Active
Error	CS1002	; expected	AndroidBinding1	C:	41	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	41	Active
Error	CS1002	; expected	AndroidBinding1	C:	41	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	58	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	70	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	82	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	95	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	107	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	122	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	134	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	149	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	161	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	173	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	186	Active
Error	CS0106	The modifier 'public' is not valid for this item	AndroidBinding1	C:	201	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	212	Active
Error	CS1022	Type or namespace definition, or end-of-file expected	AndroidBinding1	C:	213	Active

**Comment by MhAllan** [comment]
If you learn Kotlin and Swift you will finish your app before the binding is fixed, if it ever gets fixed.

**Comment by bbhxwl** [comment]
I don't know how to bind an aar, can you help me?

**Reply by MhAllan** [reply]
no good tools and endless effort, good luck

**Comment by caoyiliang** [comment]
Thank you for your letter, I will reply as soon as possible
